File: packages/react-native/ReactAndroid/src/main/java/com/facebook/react/common/MapBuilder.java
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook.react.common;

import com.facebook.infer.annotation.Nullsafe;
import java.util.HashMap;
import java.util.Map;

/**
 * Utility class for creating maps
 *
 * @deprecated Use Kotlin's built-in collections extensions
 */
@Nullsafe(Nullsafe.Mode.LOCAL)
@Deprecated
public class MapBuilder {

  /** Creates an instance of {@code HashMap} */
  public static <K, V> HashMap<K, V> newHashMap() {
    return new HashMap<K, V>();
  }

  /** Returns the empty map. */
  public static <K, V> Map<K, V> of() {
    return newHashMap();
  }

  /** Returns map containing a single entry. */
  public static <K, V> Map<K, V> of(K k1, V v1) {
    Map map = of();
    map.put(k1, v1);
    return map;
  }

  /** Returns map containing the given entries. */
  public static <K, V> Map<K, V> of(K k1, V v1, K k2, V v2) {
    Map map = of();
    map.put(k1, v1);
    map.put(k2, v2);
    return map;
  }

  /** Returns map containing the given entries. */
  public static <K, V> Map<K, V> of(K k1, V v1, K k2, V v2, K k3, V v3) {
    Map map = of();
    map.put(k1, v1);
    map.put(k2, v2);
    map.put(k3, v3);
    return map;
  }

  /** Returns map containing the given entries. */
  public static <K, V> Map<K, V> of(K k1, V v1, K k2, V v2, K k3, V v3, K k4, V v4) {
    Map map = of();
    map.put(k1, v1);
    map.put(k2, v2);
    map.put(k3, v3);
    map.put(k4, v4);
    return map;
  }

  /** Returns map containing the given entries. */
  public static <K, V> Map<K, V> of(K k1, V v1, K k2, V v2, K k3, V v3, K k4, V v4, K k5, V v5) {
    Map map = of();
    map.put(k1, v1);
    map.put(k2, v2);
    map.put(k3, v3);
    map.put(k4, v4);
    map.put(k5, v5);
    return map;
  }

  /** Returns map containing the given entries. */
  public static <K, V> Map<K, V> of(
      K k1, V v1, K k2, V v2, K k3, V v3, K k4, V v4, K k5, V v5, K k6, V v6) {
    Map map = of();
    map.put(k1, v1);
    map.put(k2, v2);
    map.put(k3, v3);
    map.put(k4, v4);
    map.put(k5, v5);
    map.put(k6, v6);
    return map;
  }

  /** Returns map containing the given entries. */
  public static <K, V> Map<K, V> of(
      K k1, V v1, K k2, V v2, K k3, V v3, K k4, V v4, K k5, V v5, K k6, V v6, K k7, V v7) {
    Map map = of();
    map.put(k1, v1);
    map.put(k2, v2);
    map.put(k3, v3);
    map.put(k4, v4);
    map.put(k5, v5);
    map.put(k6, v6);
    map.put(k7, v7);
    return map;
  }

  /** Returns map containing the given entries. */
  public static <K, V> Builder<K, V> builder() {
    return new Builder();
  }

  public static final class Builder<K, V> {

    private Map mMap;
    private boolean mUnderConstruction;

    private Builder() {
      mMap = newHashMap();
      mUnderConstruction = true;
    }

    public Builder<K, V> put(K k, V v) {
      if (!mUnderConstruction) {
        throw new IllegalStateException("Underlying map has already been built");
      }
      mMap.put(k, v);
      return this;
    }

    public Map<K, V> build() {
      if (!mUnderConstruction) {
        throw new IllegalStateException("Underlying map has already been built");
      }
      mUnderConstruction = false;
      return mMap;
    }
  }
}

File: packages/react-native/ReactAndroid/src/main/java/com/facebook/react/common/MapBuilder.kt
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook.react.common

/** Utility class for creating maps */
@Deprecated("Use Kotlin's built-in collections extensions")
public object MapBuilder {
  /** Creates an instance of `HashMap` */
  @JvmStatic public fun <K, V> newHashMap(): HashMap<K, V> = hashMapOf()

  /** Returns the empty map. */
  @JvmStatic public fun <K, V> of(): MutableMap<K, V> = newHashMap()

  /** Returns map containing a single entry. */
  @JvmStatic public fun <K, V> of(k1: K, v1: V): MutableMap<K, V> = hashMapOf(k1 to v1)

  /** Returns map containing the given entries. */
  @JvmStatic
  public fun <K, V> of(k1: K, v1: V, k2: K, v2: V): MutableMap<K, V> = hashMapOf(k1 to v1, k2 to v2)

  /** Returns map containing the given entries. */
  @JvmStatic
  public fun <K, V> of(k1: K, v1: V, k2: K, v2: V, k3: K, v3: V): MutableMap<K, V> =
      hashMapOf(k1 to v1, k2 to v2, k3 to v3)

  /** Returns map containing the given entries. */
  @JvmStatic
  public fun <K, V> of(k1: K, v1: V, k2: K, v2: V, k3: K, v3: V, k4: K, v4: V): MutableMap<K, V> =
      hashMapOf(k1 to v1, k2 to v2, k3 to v3, k4 to v4)

  /** Returns map containing the given entries. */
  @JvmStatic
  public fun <K, V> of(
      k1: K,
      v1: V,
      k2: K,
      v2: V,
      k3: K,
      v3: V,
      k4: K,
      v4: V,
      k5: K,
      v5: V
  ): MutableMap<K, V> = hashMapOf(k1 to v1, k2 to v2, k3 to v3, k4 to v4, k5 to v5)

  /** Returns map containing the given entries. */
  @JvmStatic
  public fun <K, V> of(
      k1: K,
      v1: V,
      k2: K,
      v2: V,
      k3: K,
      v3: V,
      k4: K,
      v4: V,
      k5: K,
      v5: V,
      k6: K,
      v6: V
  ): MutableMap<K, V> = hashMapOf(k1 to v1, k2 to v2, k3 to v3, k4 to v4, k5 to v5, k6 to v6)

  /** Returns map containing the given entries. */
  @JvmStatic
  public fun <K, V> of(
      k1: K,
      v1: V,
      k2: K,
      v2: V,
      k3: K,
      v3: V,
      k4: K,
      v4: V,
      k5: K,
      v5: V,
      k6: K,
      v6: V,
      k7: K,
      v7: V
  ): MutableMap<K, V> =
      hashMapOf(k1 to v1, k2 to v2, k3 to v3, k4 to v4, k5 to v5, k6 to v6, k7 to v7)

  /** Returns map containing the given entries. */
  @JvmStatic public fun <K, V> builder(): Builder<K, V> = Builder()

  public class Builder<K, V> internal constructor() {
    private val map: MutableMap<K, V> = newHashMap()
    private var underConstruction = true

    public fun put(k: K, v: V): Builder<K, V> {
      check(underConstruction) { "Underlying map has already been built" }
      map[k] = v
      return this
    }

    public fun build(): Map<K, V> {
      check(underConstruction) { "Underlying map has already been built" }
      underConstruction = false
      return map
    }
  }
}
